ENH, SIMD: Initial implementation of Highway wrapper

seiko2plus opened the pull request:
A thin wrapper over Google's Highway SIMD library to simplify its interface. This commit provides the implementation of that wrapper, consisting of:

- simd.hpp: Main header defining the SIMD namespaces and configuration
- simd.inc.hpp: Template header included multiple times with different namespaces

The wrapper eliminates Highway's class tags by:
- Using lane types directly which can be deduced from arguments
- Leveraging namespaces (np::simd and np::simd128) for different register widths

A README is included to guide usage and document design decisions.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

seiko2plus committed:
ENH, SIMD: Initial implementation of Highway wrapper

A thin wrapper over Google's Highway SIMD library to simplify its interface.
This commit provides the implementation of that wrapper, consisting of:

- simd.hpp: Main header defining the SIMD namespaces and configuration
- simd.inc.hpp: Template header included multiple times with different namespaces

The wrapper eliminates Highway's class tags by:
- Using lane types directly which can be deduced from arguments
- Leveraging namespaces (np::simd and np::simd128) for different register widths

A README is included to guide usage and document design decisions.

tylerjereddy commented on numpy/_core/src/common/simd/README.md:
```diff
@@ -0,0 +1,127 @@
+
+## Available Operations
+
+The wrapper provides the following common operations that used in NumPy, additional Highway operations can be accessed via then 
```
minor copyedits

-> that are used in 

-> via the `hn`

tylerjereddy commented on numpy/_core/src/common/simd/README.md:
```diff
@@ -0,0 +1,127 @@
+
+To add more operations from Highway:
+
+Import them in the `simd.inc.hpp` file using the `using` directive or define warpper intrinsic 
```
warpper -> wrapper?

tylerjereddy reviewed (commented):
I was morbidly curious and read the `README` so added minor copyedits.

ixgbe commented:
Hi, Could you share the expected timeline or next steps for merging this PR? Thanks!

Mousius reviewed (changes requested):
First read, will process more over time 😸

seiko2plus commented:
> I was morbidly curious and read the README so added minor copyedits.

@tylerjereddy, Thanks, I have addressed them. Feel free to give it a second round.

> Hi, Could you share the expected timeline or next steps for merging this PR? Thanks!

@ixgbe, I'm was aiming to get feedback on the approach first that outlined in your PR #28490 but so far no strong opinions against it, then we can move forward with merging. Once this PR gets merged we can follow on your PR as fast as possible. Please, let me know if you have any specific suggestions or any concerns against this approach.

> First read, will process more over time 

@Mousius, Thank you! I've updated the README to cover your questions. Please give it a second read.

ixgbe commented:
> > I was morbidly curious and read the README so added minor copyedits.
> 
> @tylerjereddy, Thanks, I have addressed them. Feel free to give it a second round.
> 
> > Hi, Could you share the expected timeline or next steps for merging this PR? Thanks!
> 
> @ixgbe, I'm was aiming to get feedback on the approach first that outlined in your PR #28490 but so far no strong opinions against it, then we can move forward with merging. Once this PR gets merged we can follow on your PR as fast as possible. Please, let me know if you have any specific suggestions or any concerns against this approach.
> 
> > First read, will process more over time
> 
> @Mousius, Thank you! I've updated the README to cover your questions. Please give it a second read.
LGTM (Looks Good To Me)! Thanks for coordinating this.

ixgbe commented on numpy/_core/src/common/simd/simd.inc.hpp:
```diff
@@ -0,0 +1,102 @@
+template <>
+constexpr bool kSupportLane<hwy::float16_t> = HWY_HAVE_FLOAT16 != 0;
+template <>
+constexpr bool kSupportLane<double> = HWY_HAVE_FLOAT64 != 0;
```
we should consider long double:
template <>constexpr bool kSupportLane<long double> = false;

seiko2plus replied:
Done but with exception in case of `sizeof(long double) == sizeof(double)`, which is the case for most ARM architecture implementations and msvc for both x86 and arm.

ixgbe commented:
when I compile numpy source code(https://github.com/numpy/numpy/pull/28608) on RVV1.0 platform, We’ve identified two pain points that require further investigation: 1、`constexpr int UNROLL = (NPY_SIMD == 128 ? 4 : 2);`   Since we have migrated from NPY_SIMD to NPY_SIMDX, the NPY_SIMD macro is now undefined for RVV (RISC-V Vector), which will trigger compile-time errors."
2、NPY_SIMD、NPY_SIMD_F32、NPY_SIMD_F64 for RISC-V Vector are undefined. The compiler error messages are shown in the figure below.      How should we resolve this? Any insights would be greatly appreciated!
![image](https://github.com/user-attachments/assets/789b8df5-4f2f-43e1-8e09-966639cf54eb)

seiko2plus committed:
SIMD: Update wrapper with improved docs and type support

  - Fix hardware/platform terminology in documentation for clarity
  - Add support for long double in template specializations
  - Add kMaxLanes constant to expose maximum vector width information
  - Follows clang formatting style for consistency with NumPy codebase.

seiko2plus commented:
@ixgbe,
I have added kMaxLanes which can be used in the first case:
```cpp
constexpr int UNROLL = (NPY_SIMD == 128 ? 4 : 2);
```
to:
```cpp
constexpr int kUnrollSize = (kMaxLanes<uint8_t> == 16 ? 4 : 2);
```

For loadable_* and storeable_*, you should avoid using them since Highway's scatter/gather intrinsics use scalar operations. However, you will need to implement another utility function to check if the stride is divisible by the size of the type.

amane-ame commented on numpy/_core/src/common/simd/simd.inc.hpp:
```diff
@@ -0,0 +1,126 @@
+}
+
+/// Returns the number of vector lanes based on the lane type.
+template <typename TLane>
+HWY_API constexpr size_t
+Lanes(TLane tag = 0)
+{
+    return hn::Lanes(_Tag<TLane>());
+}
```
`hn::Lanes` is not always constexpr, see [this](https://github.com/numpy/numpy/actions/runs/14486834354/job/40634365835?pr=28718).

seiko2plus replied:
`constexpr` shouldn't be used within the variable definition; const should be used instead to avoid build errors. I've updated the patch to use `HWY_LANES_CONSTEXPR` instead, which avoids passing `constexpr` for scalable SIMD extensions.

amane-ame replied:
This feature was introduced weeks ago: https://github.com/google/highway/commit/99bb05378be47b4c21baef4c8ff0db831a2a88f1
Please update the submodule version.

seiko2plus replied:
done, thank you.

amane-ame commented on numpy/_core/src/common/simd/simd.inc.hpp:
```diff
@@ -0,0 +1,126 @@
+ * @tparam TLane The lane type to check for support.
+ */
+template <typename TLane>
+constexpr bool kSupportLane = NPY_SIMDX != 0;
+
+#if NPY_SIMDX
+// Define lane type support based on Highway capabilities
+template <>
+constexpr bool kSupportLane<hwy::float16_t> = HWY_HAVE_FLOAT16 != 0;
+template <>
+constexpr bool kSupportLane<double> = HWY_HAVE_FLOAT64 != 0;
+template <>
+constexpr bool kSupportLane<long double> =
```
```suggestion
constexpr inline bool kSupportLane = NPY_SIMDX != 0;

#if NPY_SIMDX
// Define lane type support based on Highway capabilities
template <>
constexpr inline bool kSupportLane<hwy::float16_t> = HWY_HAVE_FLOAT16 != 0;
template <>
constexpr inline bool kSupportLane<double> = HWY_HAVE_FLOAT64 != 0;
template <>
constexpr inline bool kSupportLane<long double> =
```

seiko2plus replied:
Inline would force all translation units to share the same value, which isn't appropriate in this case. I've updated the patch to use an anonymous namespace instead.

seiko2plus committed:
SIMD: Improve isolation and constexpr handling in wrapper

  - Add anonymous namespace around implementation to ensure each
    translation unit gets its own constants based on local flags
  - Use HWY_LANES_CONSTEXPR for Lanes function to ensure proper
    constexpr evaluation across platforms

amane-ame commented on numpy/_core/src/common/simd/simd.inc.hpp:
```diff
@@ -0,0 +1,132 @@
+using hn::Sub;
+using hn::Xor;
+
+} // namespace anonymous
+
+#endif  // NPY_SIMDX
```
```suggestion
#endif  // NPY_SIMDX

} // namespace anonymous
```

seiko2plus replied:
my bad, thank you.

seiko2plus committed:
Update Highway submodule to latest master

seiko2plus committed:
SIMD: Fix compile error by using MaxLanes instead of Lanes for array size

  Replace hn::Lanes(f64) with hn::MaxLanes(f64) when defining the index array
  size to fix error C2131: "expression did not evaluate to a constant". This error
  occurs because Lanes() isn't always constexpr compatible, especially with
  scalable vector extensions. MaxLanes() provides a compile-time constant value
  suitable for static array allocation and should be used with non-scalable SIMD
  extensions when defining fixed-size arrays.

seiko2plus commented on numpy/_core/src/common/simd/README.md:
```diff
@@ -0,0 +1,127 @@
+   - This prevents unexpected behavior when optimization is disabled
+
+2. **Namespace Design**
+   - `np::simd`: Maximum width SIMD operations (scalable)
+   - `np::simd128`: Fixed 128-bit SIMD operations
```
Done, please check `Why Namespaces Are Essential for This Design?`  and `Why Namespaces and Why Not Just Use Highway Directly?`.

seiko2plus commented on numpy/_core/src/common/simd/simd.hpp:
```diff
@@ -0,0 +1,80 @@
+ */
+#define NPY_SIMDX (HWY_TARGET != HWY_SCALAR)
+
+// Indicates if the SIMD operations are available for float16.
+#define NPY_SIMDX_F16 (NPY_SIMDX && HWY_HAVE_FLOAT16)
+// Note: Highway requires SIMD extentions with native float32 support, so we don't need
+// to check for it.
+
+// Indicates if the SIMD operations are available for float64.
+#define NPY_SIMDX_F64 (NPY_SIMDX && HWY_HAVE_FLOAT64)
+
+// Indicates if the SIMD floating operations are natively supports fma.
+#define NPY_SIMDX_FMA (NPY_SIMDX && HWY_NATIVE_FMA)
```
We can use HWY variants, but under `#if NPY_SIMDX`, please check design notes `3. Feature Detection Constants vs. Highway Constants`

r-devulap commented on numpy/_core/src/common/simd/simd.inc.hpp:
```diff
@@ -0,0 +1,132 @@
+
+/// Unaligned load of a vector from memory.
+template <typename TLane>
+HWY_API Vec<TLane>
```
Do these template functions not require `HWY_ATTR`?

seiko2plus replied:
`HWY_API` implies both `HWY_ATTR` and `HWY_INLINE`. However, since we use source-based CPU dispatching, there's no need for `HWY_ATTR`, but I used `HWY_API` in case we need to move it to Highway later.

r-devulap reviewed (changes requested):
Minor comments but mostly LGTM. I still need to wrap my head around it and will give it a second read tomorrow.

seiko2plus commented on numpy/_core/src/common/simd/simd.hpp:
```diff
@@ -0,0 +1,80 @@
+ *
+ * Therefore, we only enable SIMD when the Highway target is not scalar.
+ */
+#define NPY_SIMDX (HWY_TARGET != HWY_SCALAR)
```
> Could we rename NPY_SIMDX* to NPY_HWY_SIMD*? It makes it more explicit that it is a combination of NPY and HWY macros and also makes it easy to search for in the source code.

`NPY_HWY_SIMD_*` is a bit long. Maybe `NPY_HWY_*` would be better. I chose the original naming based on the namespace `np::simd`.

> Also, NPY_SIMD_F32 is used in a lot of places. Perhaps to make it consistent we could define a macro for float32 as well?

No need for it. It should be replaced with `NPY_SIMDX` as Highway requires SIMD extensions with float32 support.

r-devulap replied:
`NPY_HWY_*` sounds good to me.

seiko2plus commented on numpy/_core/src/common/simd/simd.hpp:
```diff
@@ -0,0 +1,80 @@
+
+// Indicates if the SIMD operations are available for float16.
+#define NPY_SIMDX_F16 (NPY_SIMDX && HWY_HAVE_FLOAT16)
+// Note: Highway requires SIMD extentions with native float32 support, so we don't need
```
They assumed that all SIMD extensions support float32, as they didn't support z/Architecture (zsystem) from the beginning. While ARMv7 doesn't support float64, they had to add specific definitions for it. The only exception is z13's "vx" instruction set, which doesn't support float32. Highway's z/Architecture support starts from z14 onwards.

seiko2plus commented on numpy/_core/src/common/simd/simd.inc.hpp:
```diff
@@ -0,0 +1,132 @@
+
+/// Maximum number of lanes supported by the SIMD extension for the specified lane type.
+template <typename TLane>
+constexpr size_t kMaxLanes = HWY_MAX_LANES_D(_Tag<TLane>);
```
indeed:
https://github.com/numpy/numpy/blob/c7884f0f7a99684671ada0dec597c3b0f4d43e56/numpy/_core/src/common/simd/simd.hpp#L55-L75

r-devulap reviewed (commented):
I have had some time to sit on this a little bit and here are my thoughts. As far as I understand, the main goal of having a wrapper around highway intrinsics is to get rid of the tags that highway ops require. For example, this piece of code

```
namespace hn = hwy::HWY_NAMESPACE;
template<typename T>
HWY_ATTR void add(T* src, T* dst, size_t N) {
    const hn::ScalableTag<T> ftype;
    hn::Vec<decltype(ftype)> v = hn::LoadU(ftype, src);
    v = hn::Add(v, v);
    hn::StoreU(v, ftype, dst);
}
```

can be written without tags: 

```
using namespace np::simd;
template<typename T>
void add(T* src, T* dst, size_t N) {
    Vec<T> v = LoadU(src);
    v = Add(v, v);
    StoreU(v, dst);
}
```

I do agree that the latter way looks cleaner and avoids having to redefine tags and highway vector for every kernel (like here: https://github.com/numpy/numpy/pull/28490#discussion_r2017760575), I do have to point out couple of downsides of having this wrapper: 

1) This adds an extra layer of abstraction for the intrinsics that needs to be maintained and potentially add an extra layer to debug.
2) If we want all our highway kernels to be tag free, then this wrapper needs to be expanded with time for all highway ops used which is extra work. 

For example, if we want the kernel defined in `loops_hyperbolic.dispatch.cpp.src` to be tags free, the wrapper needs to expanded to include all these highway ops: 

```
hn::InterleaveLower
hn::TwoTablesLookupLanes
hn::RebindMask
hn::GatherIndex
hn::ConcatLowerLower
hn::BitCast
hn::Sub
hn::Mul
```

While the approach proposed here avoids explicit tags in Highway operations, the added abstraction layer introduces maintenance overhead without significant benefits. Directly using Highway's native tag syntax (e.g., `hn::LoadU(ftype, src)`) appears sufficiently readable and not hard to write. 

Here is my proposed alternative: 

1) Centralize type definitions in simd.hpp with predefined tags/vectors
2) Reuse these definitions across all Highway kernels
3) Avoid op-specific expansions through generic implementations

This way we avoid introducing an extra abstraction layer and eliminate the need to expand wrappers for every Highway operation.

@seiko2plus Do you think this would be a simpler and more maintainable solution? ping @jan-wassenberg and @Mousius for their opinions.

seiko2plus commented:
@r-devulap,

I appreciate your thoughtful analysis. You make valid points about the trade-offs involved with this wrapper approach.
The primary motivation for this wrapper is indeed to make SIMD code more approachable by eliminating the tag-based syntax, because the current way of Highway seems to encourages users to produce more complicated meta-programming code in their final kernels.

Also, modern C++17 features allow us to introduce simpler constants to deal with SFINAE with cleaner code: e.g., `kMaxLanes<float>` rather than `HWY_MAX_LANES_D<ScalableTag<float>>` or `kSupportLane<double>` rather than `#ifdef HWY_HAVE_FLOAT64`.


> This adds an extra layer of abstraction for the intrinsics that needs to be maintained and potentially add an extra layer to debug.


This wrapper is exceptionally thin. Most functions are simple one-liners that just forward to Highway with the appropriate tag. There's very little to maintain or debug in the wrapper itself. If an issue occurs, the debugging process is straightforward - the problem will either be in the calling code or in Highway itself, with the wrapper simply passing values through. The mapping between wrapper functions and Highway functions is direct and transparent. See:
https://github.com/numpy/numpy/blob/c7884f0f7a99684671ada0dec597c3b0f4d43e56/numpy/_core/src/common/simd/simd.inc.hpp#L47-L52


> If we want all our highway kernels to be tag free, then this wrapper needs to be expanded with time for all highway ops used which is extra work.


Not all Highway operations need to be functionally wrapped. Most of Highway operations are already tag-free; extra work is nothing compared to the benefit of having cleaner kernels. Developers can still access the full Highway API through the `hn` namespace alias that's already included in the wrapper before deciding to add a wrapper for a specific operation. Do you think this expansion process is a burden?
https://github.com/numpy/numpy/blob/c7884f0f7a99684671ada0dec597c3b0f4d43e56/numpy/_core/src/common/simd/README.md?plain=1#L142-L157

> While the approach proposed here avoids explicit tags in Highway operations, the added abstraction layer introduces maintenance overhead without significant benefits.

The benefits are clear: code becomes more concise, more readable, less prone to errors and encourages users to produce simpler code. As we decided to replace C template sources `.c.src` with C++ code for simplicity, would it be wise to choose a cleaner approach even if that requires a thin wrapper to Highway that potentially will be moved to Highway itself in the end?

> Here is my proposed alternative:
> Centralize type definitions in simd.hpp with predefined tags/vectors
> Reuse these definitions across all Highway kernels
> Avoid op-specific expansions through generic implementations
> This way we avoid introducing an extra abstraction layer and eliminate the need to expand wrappers for every Highway operation.


Not sure what the point of predefining tags is? Why not just `hn::ScalableTag<float>{}`? or `template<typename T> constexpr auto kScalable = hn::ScalableTag<T>{};` with `template <typename T> using ScalableVec = hn::Vec<hn::ScalableTag<T>>;?`

Using your previous example, your suggestion would be like this:

```C++
template<typename T>
void add(T* src, T* dst, size_t N) {
    // ScalableVec<T> v = hn::LoadU(kScalable<T>, src);
    // bad practice to use auto everywhere but can we tolerate it?
    auto v = hn::LoadU(kScalable<T>, src);
    v = hn::Add(v, v);
    hn::StoreU(v, kScalable<T>, dst);
}
```
How about `kMaxLanes`, `kSupportLane`, and constants `NPY_SIMDX_*`? Based on your alternative, are we going to define them in simd.hpp too?


> Do you think this would be a simpler and more maintainable solution?

I'm more concerned about maintaining the kernels than the wrapper itself.
With your suggestion, we still create some form of abstraction - just to avoid simple one-liner wrappers where needed.
All code involves abstraction; the question is at what level. At NumPy's scale, the thin wrapper approach is absolutely worth it.



> There was also a concern that tags complicate python bindings because many more wrapper functions would have to be generated there - one for half, quarter vectors etc.
> I suppose we can for purposes of that python binding only export the full-vector version that the currently proposed wrapper functions here would provide.


@jan-wassenberg, Thank you for your input. If we moved the wrapper to Highway to replace `google-tests` with `pytest`, it would still be possible to use `hwy::scalable::<HWY_TARGETS>`, `hwy::fixed128::<TARGETS>`, etc. in the bindings.
`pytest` is dynamic enough to handle this.

jan-wassenberg commented:
One quick comment:

> // bad practice to use auto everywhere but can we tolerate it?

I'd advise against using auto everywhere. Kernels that involve range reduction may involve multiple types (at least int32 and f32, possibly more), and I think it's helpful to see the type mentioned in the code.

Mousius commented:
> This wrapper is exceptionally thin. All functions are simple one-liners that just forward to Highway with the appropriate tag. There's very little to maintain or debug in the wrapper itself.

This is potentially my most pressing question: Will it continue to be thin? It seems we'll have a lot of thin wrappers.

@jan-wassenberg do you see issues with these being contributed directly to Highway? I'd imagine a lot of the C++17 stuff is a welcome addition to the macro variants?

I'm not seeing anything that is NumPy-centric or needs to be NumPy-centric just yet, and it appears we already know roughly what we'd like to contribute.

> Not all Highway operations need to be functionally wrapped. Most of the Highway operations are already tag-free; extra work is nothing compared to the benefit of having cleaner kernels. 

I'd also suggest that the extra work may not be necessary if it's this minimal for users. I prefer @r-devulap's ideas of centralised tags for NumPy types and explicit typing in the SIMD kernels.

> The benefits are clear: code becomes more concise, more readable, less prone to errors and encourages users to produce simpler code. 

This could be debated. As @jan-wassenberg has mentioned above, it is valuable to see the types and conversions explicitly in SIMD kernels to avoid errors. The code itself may contain fewer characters, but that doesn't necessarily correlate with being easier to read. Personally, I've found the verbosity of the SIMD kernels justified due to their details.

seiko2plus commented:
@jan-wassenberg,

> I'd advise against using auto everywhere. Kernels that involve range reduction may involve multiple types (at least int32 and f32, possibly more), and I think it's helpful to see the type mentioned in the code.

Totally agree with you as mentioned "bad practice". Based on @r-devulap's idea of using centralized tags, the most simplified example would be like:
```C++
template<typename T>
void add(T* src, T* dst, size_t N) {
    ScalableVec<T> v = hn::LoadU(kScalable<T>, src);
    v = hn::Add(v, v);
    hn::StoreU(v, kScalable<T>, dst);
}
````

And that's why I believe simplifying Highway will encourages users to produce more clearer and simple code.

@Mousius,

> Will it continue to be thin? It seems we'll have a lot of thin wrappers.

Yes definitely. We're just erasing tags and using modern C++ features. We're in 2025 - if we have a chance to make our meta-programming much simpler by leveraging modern C++ features, we should take it. Highway had legacy reasons to support C++11, but we don't.

> I'm not seeing anything that is NumPy-centric or needs to be NumPy-centric just yet, and it appears we already know roughly what we'd like to contribute.

I agree that most of this wrapper isn't NumPy-specific and could benefit the broader Highway ecosystem. That's why it would be great to move it to Highway as we agreed before.

> This could be debated. As @jan-wassenberg has mentioned above, it is valuable to see the types and conversions explicitly in SIMD kernels to avoid errors. The code itself may contain fewer characters, but that doesn't necessarily correlate with being easier to read. Personally, I've found the verbosity of the SIMD kernels justified due to their details.

I think you misunderstood my comment https://github.com/numpy/numpy/pull/28622#issuecomment-2837251234. Let me explain with more details:

@r-devulap suggested using centralized tags, and my response was "there's no need for them, what's the point of having them?" Instead, we can use template variables (introduced in C++14) rather than explicitly defining each tag (if we decide to tolerate tags).

So instead of explicitly defining each tag:
```C++
const hn::ScalableTag<float> f32;
const hn::ScalableTag<int32_t> s32;
```
we can define a template variable like this:
```C++
template<typename T> constexpr auto kScalable = hn::ScalableTag<T>{};
```
And by relying on the lane type, you can define any tag simply like this:
```C++
kScalable<float>; // equivalent to f32
kScalable<int32_t>; // equivalent to s32
````

Mousius commented:
> I agree that most of this wrapper isn't NumPy-specific and could benefit the broader Highway ecosystem. That's why it would be great to move it to Highway as we agreed before.

I'm sorry, my question was probably misformed. What prevents us from contributing this directly to Highway now?

> Instead, we can use template variables (introduced in C++14) rather than explicitly defining each tag (if we decide to tolerate tags).

Couldn't we also use `ScalableTag<float>` everywhere, which is two characters more than `kScalable<float>`?

seiko2plus commented:
> I personally think doesn't add much value

It's different preferences, I guess. To me, counting on ADL and being able to force change SIMD width with just a single line when needed is a great advantage.

> Thin or not, it is still additional code

But we are going to transferring the additional code from the kernels to the wrapper. 

> What prevents us from contributing this directly to Highway now?

Well, once it becomes mature enough we can pass it. New ideas may emerge anytime - why should we worry about backward compatibility at this stage when other projects aren't yet depending on it?

> Couldn't we also use ScalableTag<float> everywhere, which is two characters more than kScalable<float>?

`ScalableTag` is the type, while `kScalable` would be the tag instance. Yes, we could use the type directly with extra characters `()` or `{}`.

r-devulap commented:
> It's different preferences, I guess. To me, counting on ADL and being able to force change SIMD width with just a single line when needed is a great advantage.

@seiko2plus if you find it advantageous and feel strongly about it, I don't mind putting this in. As you said in an earlier comment: developers can still access the full Highway API through the `hn namespace` alias.

r-devulap commented:
@seiko2plus Can we make this change before merging? https://github.com/numpy/numpy/pull/28622/files#r2057043403

Mousius commented:
Discussing in the Optimisation team meeting:

* `#define NPY_SIMDX (HWY_TARGET != HWY_SCALAR)` will always be `True` as Highway is going to default `EMU128` which will allow for compiler auto-vectorisation by default. Therefore `NPY_SIMDX` is likely unnecessary?
* Highway is happy to host the wrapper to remove lane tags under contrib given it's fairly clear why it exists and why it's useful, which would save us hosting it in NumPy. We should contribute it directly under `contrib` for Highway which equally allows for it to evolve further if necessary?

seiko2plus committed:
SIMD: Rename NPY_SIMDX to NPY_HWY

Rename Highway wrapper macros for clarity:
- NPY_SIMDX → NPY_HWY
- NPY_SIMDX_F16 → NPY_HWY_F16
- NPY_SIMDX_F64 → NPY_HWY_F64
- NPY_SIMDX_FMA → NPY_HWY_FMA

To avoids confusion with legacy SIMD macros.

seiko2plus commented:
> @seiko2plus Can we make this change before merging? https://github.com/numpy/numpy/pull/28622/files#r2057043403

Done.

> #define NPY_SIMDX (HWY_TARGET != HWY_SCALAR) will always be True as Highway is going to default EMU128 which will allow for compiler auto-vectorisation by default. Therefore NPY_SIMDX is likely unnecessary?

`NPY_HWY (NPY_SIMDX)` is needed for two important reasons: to completely disable optimization via disable-optimization and to clear the baseline features via `cpu-baseline=none`. This is important because Highway will still enable `SSE3` by default even if `SSE` flags aren't provided, since `SSE3` is enabled by default on `gcc/clang`. By the way, `Highway` doesn't correctly emulate many `SSE41` SIMD functions, but this isn't a concern since PR #28896 will raise the baseline to x86-64-v2 (level 2).

> Highway is happy to host the wrapper to remove lane tags under contrib given it's fairly clear why it exists and why it's useful, which would save us hosting it in NumPy. We should contribute it directly under contrib for Highway which equally allows for it to evolve further if necessary?

As I mentioned before, I'd prefer to use it within NumPy first rather than moving it to Highway until it's mature enough. Hosting in NumPy will make it easier to make changes as we refine the implementation.

seiko2plus committed:
Disable Highway EMU128 scalar emulation

Skip Highway's EMU128 in favor of NumPy's scalar implementations for
due to strict IEEE 754 floating-point compliance requirements

seiko2plus commented:
> I believe this is already possible by only enabling EMU128? I don't see why we need NPY_SIMDX for this.

When the option `disable-optimization` is set to true, we should disable all CPU optimized code including dispatch, SIMD, and loop unrolling.
After a quick look at the `EMU128` implementation, I realized that we should disable it too to avoid any kind of fast math operations. Also, compiling with `fp-strict` may not lead to auto-vectorization and may end up with register spilling. Lets stay safe for now, till we get done from moving to Highway.

Mousius commented:
> When the option `disable-optimization` is set to true, we should disable all CPU optimized code including dispatch, SIMD, and loop unrolling.
> After a quick look at the `EMU128` implementation, I realized that we should disable it too to avoid any kind of fast math operations. Also, compiling with `fp-strict` may not lead to auto-vectorization and may end up with register spilling. Lets stay safe for now, till we get done from moving to Highway.

We already have `NPY_DISABLE_OPTIMIZATION` to completely disable the optimised kernels?

I'm not sure whether a lack of optimisation in `EMU128` is hugely worrying given it's the fallback.

r-devulap commented:
Thanks @seiko2plus. I'm okay with merging it with the assumption that we will move this into Highway sooner rather than later.

r-devulap reviewed (approved):
LGTM.

r-devulap commented:
Looks like I cant merge this until @Mousius approves or his review is dismissed.

Mousius commented:
Thanks for your patience @seiko2plus!
